Add Omnity Hub

chezhe opened the pull request:
**NOTE**

#### Please enable "Allow edits by maintainers" while putting up the PR.

---

> - If you would like to add a `volume/fees/revenue` adapter please submit the PR [here](https://github.com/DefiLlama/adapters).
> - If you would like to add a `liquidations` adapter, please refer to [this readme document](https://github.com/DefiLlama/DefiLlama-Adapters/tree/main/liquidations) for details.

1. Once your adapter has been merged, it takes time to show on the UI. If more than 24 hours have passed, please let us know in Discord.
2. Sorry, We no longer accept fetch adapter for new projects, we prefer the tvl to computed from blockchain data, if you have trouble with creating a the adapter, please hop onto our discord, we are happy to assist you.
3. Please fill the form below  **only if the PR is for listing a new protocol** else it can be ignored/replaced with reason/details about the PR
4. **For updating listing info** It is a different repo, you can find your listing in this file: https://github.com/DefiLlama/defillama-server/blob/master/defi/src/protocols/data2.ts, you can  edit it there and put up a PR
5. Do not edit/push `package-lock.json` file as part of your changes, we use lockfileVersion 2, and most use v1 and using that messes up our CI
6. No need to go to our discord and announce that you've created a PR, we monitor all PRs and will review it asap

---
##### Name (to be shown on DefiLlama): Omnity Hub


##### Twitter Link: https://x.com/OmnityNetwork


##### List of audit links if any: http://omnity.network/blocksec_omnity_v1.0_signed.pdf

##### Website Link: https://hub.omnity.network/


##### Logo (High resolution, will be shown with rounded borders): https://hub.omnity.network/logo.svg


##### Current TVL: 366,376,702


##### Treasury Addresses (if the protocol has treasury)


##### Chain: ICP, Osmosis


##### Coingecko ID (so your TVL can appear on Coingecko, leave empty if not listed): (https://api.coingecko.com/api/v3/coins/list)


##### Coinmarketcap ID (so your TVL can appear on Coinmarketcap, leave empty if not listed): (https://api.coinmarketcap.com/data-api/v3/map/all?listing_status=active,inactive,untracked&start=1&limit=10000)


##### Short Description (to be shown on DefiLlama): Seamlessly connect Bitcoin to L1s and L2s with a trust-minimized, fully on-chain architecture


##### Token address and ticker if any:


##### Category (full list at  https://defillama.com/categories) *Please choose only one: Bridge


##### Oracle Provider(s): Specify the oracle(s) used (e.g., Chainlink, Band, API3, TWAP, etc.):
##### Implementation Details: Briefly describe how the oracle is integrated into your project:
##### Documentation/Proof: Provide links to documentation or any other resources that verify the oracle's usage:

##### forkedFrom (Does your project originate from another project):


##### methodology (what is being counted as tvl, how is tvl being calculated): The TVL of Omnity Hub refers to the total value of assets locked by users on bridge addresses or smart contracts of the settlement chain (the asset issuance chain) to enable cross-chain operations.
Omnity Hub adopts a hub-and-spoke architecture, in which the hub canister (an ICP smart contract) maintains the master ledger of cross-chain assets.
By querying the hub canister to obtain the token list and locked amounts, and then combining these with prices from the largest on-chain marketplace or launchpad for each token, the TVL is calculated.


##### Github org/user (Optional, if your code is open source, we can track activity): https://github.com/octopus-network

waynebruce0x commented:
What does this TVL value consist of? Is the TVL endpoint used open source?

waynebruce0x commented:
If the funds are bridged from Bitcoin to Dfinity, and the runes are locked on Bitcoin, the TVL should be counted on bitcoin. We have lots of helper functions in the repo for counting this on chain. 

How about the osmosis TVL? We cant use this endpoint if its not open source, its too opaque

llamabutler commented:
The adapter at projects/omnity-hub exports TVL: 
        
 
 
    bitcoin                   360.45 M
    
    total                    360.45 M

chezhe commented:
Bitcoin does not have smart contracts. Therefore, whenever a user wants to bridge Bitcoin Runes to another chain, Omnity generates, on demand, an asset lock address for each target chain and receiving address combination. The key for this address is the TSS Chain Key derived by the Bitcoin connector smart contract, known as the Bitcoin Customs canister.

When Runes need to be bridged back to Bitcoin, the Bitcoin Customs canister will select and spend the UTXO containing the corresponding assets, and consolidate the change assets. As a result, the set of Bitcoin asset lock addresses belonging to the Omnity system is constantly changing, and only the Bitcoin Customs canister has complete information. Therefore, it is reasonable for the DefiLlama adapter to obtain the amount of locked assets from the built-in HTTP service of the Bitcoin Customs canister.

I changed the endpoint, the locked assets are fetched from canister now, and the price is fetched from our endpoint, it's open source here https://github.com/octopus-network/api.omnity.network/blob/main/src/app/api/price/tokens/route.ts which caches the price from [maestro](https://www.gomaestro.org/) and odin.fun

I changed the chain to bitcoin, the source network of runes is bitcoin, so it should be bitcoin rather than icp, right?
